Add Redox support

jackpot51 opened the pull request:
This adds support for Redox. I will annotate changes in a GitHub review.

jackpot51 commented on Cargo.toml:
```diff
@@ -262,7 +262,7 @@ name = "ring"
 untrusted = "0.3.2"
 
-[target.'cfg(unix)'.dependencies]
+[target.'cfg(any(target_os = "redox", unix))'.dependencies]
```
Redox also needs lazy_static

jackpot51 commented on mk/top_of_makefile.mk:
```diff
@@ -41,11 +41,11 @@ else ifeq ($(findstring linux-android,$(TARGET_VENDOR)-$(TARGET_SYS)),linux-andr
 TARGET_VENDOR = unknown
 TARGET_SYS = linux
-else
+else ifeq ($(findstring redox,$(TARGET_SYS)),redox)
```
Redox is a supported system, so do not throw an error on finding it

briansmith replied:
When double-checking this before merging this, I realized that I don't understand this. What are the target triples for Redox, and in particular isn't the target triple always of the form $(architecture)-$(vendor)-$(os)-$(abi)?, e.g. "i686-unknown-redox-gnu"? If so, what's the point of this change? If not, what are the target triples for redox?

jackpot51 commented on mk/top_of_makefile.mk:
```diff
@@ -231,7 +230,9 @@ CPPFLAGS += \
 # The launchpad.net arm-none-eabi-gcc toolchain (at least) uses -fshort-enums.
 ifneq ($(TARGET_SYS),none)
-CPPFLAGS += -fstack-protector
+	ifneq ($(findstring redox,$(TARGET_SYS)),redox)
```
Redox does not support libssp at the moment

briansmith replied:
Is there a Redox issue that we could link to in a comment? Or if not, could you create one and then link it?

jackpot51 replied:
Yes, there is an issue here: https://github.com/redox-os/redox/issues/815

briansmith replied:
OK, please add a link to that issue as a comment here.

briansmith replied:
Please add a comment with a link to that issue here.

briansmith replied:
Again, similar question. Why are we doing:
```make
ifneq ($(findstring redox,$(TARGET_SYS)),redox)
```
instead of:
```make
ifneq ($(TARGET_SYS),redox)
```

We need to know this so that we can do the right thing when we replace the makefile.

jackpot51 commented on src/lib.rs:
```diff
@@ -99,7 +99,8 @@
 extern crate test as bench;
 
-#[cfg(any(all(unix,
+#[cfg(any(target_os = "redox",
```
Redox will need lazy_static (at line 196)

jackpot51 commented on src/rand.rs:
```diff
@@ -140,6 +140,28 @@ mod sysrand {
 }
 
 // Keep the `cfg` conditions in sync with the conditions in lib.rs.
+#[cfg(target_os = "redox")]
```
Redox has what Linux has at `/dev/urandom`, at `rand:`.

briansmith replied:
Rather than duplicate the code for the `urandom` submodule, let's abstract out the one difference, the path to the random file, and share the rest of the code.

jackpot51 replied:
I will see about that.

briansmith commented:
Also, is the "C" calling convention in Redox the same as Linux or different in any way? This is important for understanding whether the assembly language stuff is correct.

jackpot51 commented:
@briansmith The calling conventions are identical to those in Linux - System V ABI

I will update the documentation.

There is no plan for a getrandom syscall. Redox is a microkernel, and randomness has been sourced through a userspace process, through a device node at `rand:`. A special case for randomness *might* be made to allow for centralized entropy collection, I have an issue here: https://github.com/redox-os/redox/issues/816

It would be possible but messy to run the tests on Redox. It would involve cross compiling the tests, running Redox in QEMU, and running a script in Redox to drive the automatic download and running of the tests.

jackpot51 commented on src/rand.rs:
```diff
@@ -140,17 +143,22 @@ mod sysrand {
     use error;
 
     pub fn fill(dest: &mut [u8]) -> Result<(), error::Unspecified> {
+        #[cfg(target_os = "redox")]
```
I simplified the differences between Redox and Posix-like systems

jackpot51 commented on src/rand.rs:
```diff
@@ -57,6 +57,9 @@ pub trait SecureRandom {
 /// implemented by reading from `/dev/urandom`. (This is something that should
 /// be improved, at least for platforms that offer something better.)
 ///
+/// On Redox, `fill()` is implemented by reading from `rand:`. In the future,
```
I added the requested module-level documentation

jackpot51 commented:
@briansmith What do you think of my updates?

jackpot51 commented:
@briansmith I have squashed it, rebased, and added the contributors statement.

All I ask is that you please `cargo publish` as soon as possible, with the current 0.6* version, so that it can be used in the wild.

briansmith commented:
In 0.6.2:
8a0976b363e41ae9586391bd7da493b89862300d
429f5c25baecaccbbe649fa0ed906435bdda7754.

I tweaked your commit. Hopefully I didn't mess it up.

jackpot51 commented:
Looking good - thanks @briansmith
